process/interceptors/factory/eligibleNodesCache.go

```diff
@@ -0,0 +1,102 @@
+package factory
+
+import (
+	"sync"
+
+	"github.com/multiversx/mx-chain-core-go/core"
+	"github.com/multiversx/mx-chain-core-go/core/check"
+	"github.com/multiversx/mx-chain-go/process"
+)
+
+type eligibleNodesCache struct {
+	// eligibleNodesMap holds a map with key shard id and value
+	//		a map with key epoch and value
+	//		a map with key eligible node pk and value struct{}
+	// this is needed because:
+	// 	- a shard node will intercept proof for self shard and meta
+	// 	- a meta node will intercept proofs from all shards
+	eligibleNodesMap    map[uint32]map[uint32]map[string]struct{}
+	mutEligibleNodesMap sync.RWMutex
+
+	peerShardMapper  process.PeerShardMapper
+	nodesCoordinator process.NodesCoordinator
+}
+
+func newEligibleNodesCache(
+	peerShardMapper process.PeerShardMapper,
+	nodesCoordinator process.NodesCoordinator,
+) (*eligibleNodesCache, error) {
+	if check.IfNil(peerShardMapper) {
+		return nil, process.ErrNilPeerShardMapper
+	}
+	if check.IfNil(nodesCoordinator) {
+		return nil, process.ErrNilNodesCoordinator
+	}
+
+	return &eligibleNodesCache{
+		eligibleNodesMap: map[uint32]map[uint32]map[string]struct{}{},
+		peerShardMapper:  peerShardMapper,
+		nodesCoordinator: nodesCoordinator,
+	}, nil
+}
+
+// IsPeerEligible returns true if the provided peer is eligible
+func (cache *eligibleNodesCache) IsPeerEligible(pid core.PeerID, shard uint32, epoch uint32) bool {
+	cache.mutEligibleNodesMap.Lock()
+	defer cache.mutEligibleNodesMap.Unlock()
+
+	var err error
+	eligibleNodesForShard, hasShardCached := cache.eligibleNodesMap[shard]
+	if !hasShardCached {
+		// if no eligible list is cached for the current shard, get it from nodesCoordinator
+		eligibleNodesForShardInEpoch, err := cache.getEligibleNodesForShardInEpoch(epoch, shard)
+		if err != nil {
+			return false
+		}
+
+		eligibleNodesForShard = map[uint32]map[string]struct{}{
+			epoch: eligibleNodesForShardInEpoch,
+		}
+		cache.eligibleNodesMap[shard] = eligibleNodesForShard
+	}
+
+	eligibleNodesForShardInEpoch, hasEpochCachedForShard := eligibleNodesForShard[epoch]
+	if !hasEpochCachedForShard {
+		// get the eligible list for the new epoch from nodesCoordinator
+		eligibleNodesForShardInEpoch, err = cache.getEligibleNodesForShardInEpoch(epoch, shard)
+		if err != nil {
+			return false
+		}
+
+		// if no eligible list was cached for this epoch, it means that it is a new one
+		// it is safe to clean the cached eligible list for other epochs for this shard
+		cache.eligibleNodesMap[shard] = map[uint32]map[string]struct{}{
```

Comment:
why not cache.eligibleNodesMap[shard][epoch] = eligibleNodesForShardInEpoch?

Reply:
is this in order to cleanup also the other epochs data? 
I think in case of syncing node in previous epoch-1, you will do cleanups of epoch data back and every round, replacing epoch with epoch -1 data back and forth.

I think we should keep at least a few epochs config (like we do in nodes coordinator)

Reply:
indeed that was the reason, to delete old epochs.. enhanced the implementation with a epochsDelta of 3